doc/designs/fips_deferred_tests.md

```diff
@@ -0,0 +1,198 @@
+Deferred FIPS Self-Tests
+========================
+
+Introduction
+------------
+
+This document outlines a design to change how and when FIPS Known Answer
+Tests (KATs) and other algorithm self-tests are executed within the OpenSSL
+FIPS provider. The goal is to move from the current model, where all
+self-tests are run unconditionally at provider load time, to a deferred
+model where each test is run on-demand, the first time a specific
+cryptographic algorithm is requested for use.
+
+Background
+----------
+
+Currently, when the FIPS provider is loaded into the library, it executes
+all of the self-tests for all approved algorithms it implements. The
+power-on self-tests ensure the module is in a valid state before any
+cryptographic operations are performed.
+
+While robust, this approach has a significant drawback: it can introduce a
+noticeable latency at application startup, even if the application only
+intends to use a small subset of the available FIPS algorithms. For many
+applications, this startup cost is undesirable.
+
+The proposed design shifts to a "deferred" self-test execution model. Each
+algorithm's self-test will be run only when that algorithm is first requested.
+Once a test passes, it will not be run again. This approach aims to minimize
+startup latency while maintaining FIPS compliance. FIPS 140-3 IG 10.3.A
+"Cryptographic Algorithm Self-Test Requirements" allows for deferring KATs
+until the first invocation of the algorithm.
+
+The core module integrity check and certain essential startup tests will
```

Comment:
Note that the integrity test is a completely different class of FIPS 140-3 self-tests ("pre-operational" rather than "conditional").